File: tools/grammar/driver.cpp
/* The C++ half of the byte-identical gate: the same walk as driver.py,
 * through rakupp_grammar.hpp (header-only over librakupp). Run by
 * grammar-smoke.raku; outputs are compared byte for byte against the Raku
 * reference driver.
 *
 *   driver-cpp <grammar-file> <input-file>
 *
 * (No library path argument: the driver is LINKED against librakupp — the
 * C++ story — where the ctypes/FFI hosts dlopen it.)
 */
#include "grammar.hpp"

#include <fstream>
#include <iostream>
#include <sstream>

static std::string canon(const rakupp::Tree& t) {
    if (t.is_list()) {
        std::string out = "[";
        bool first = true;
        for (auto& e : t.list()) { if (!first) out += ","; first = false; out += canon(e); }
        return out + "]";
    }
    if (t.is_map()) { /* std::map iterates sorted, matching sorted() in the twins */
        std::string out = "{";
        bool first = true;
        for (auto& kv : t.map()) { if (!first) out += ","; first = false; out += kv.first + ":" + canon(kv.second); }
        return out + "}";
    }
    if (t.is_null()) return "";
    if (std::holds_alternative<long long>(t.v)) return std::to_string(t.int_());
    if (std::holds_alternative<bool>(t.v)) return std::get<bool>(t.v) ? "True" : "False";
    return t.str();
}

static std::string slurp(const char* path) {
    std::ifstream in(path, std::ios::binary);
    std::stringstream ss;
    ss << in.rdbuf();
    return ss.str();
}

int main(int argc, char** argv) {
    if (argc < 3) { std::cerr << "usage: driver-cpp <grammar> <input>\n"; return 2; }

    auto g = rakupp::Grammar::from_file(argv[1], "Log", "LogActions");
    auto m = g.parse(slurp(argv[2]));
    if (!m) { std::cerr << "gate: the log corpus did not parse\n"; return 1; }

    auto lines = (*m)["line"];
    size_t n = lines.size();
    std::cout << "lines " << n << "\n";
    std::cout << "islist " << (lines.is_list() ? "True" : "False") << "\n";

    for (size_t i = 0; i < n; i++)
        std::cout << lines[i]["ip"].str() << " " << lines[i]["status"].str() << "\n";

    std::cout << "made " << canon((*m)["line"][0]["size"].made()) << "\n";
    std::cout << "req.str " << (*m)["line"][42]["req"].str() << "\n";
    std::cout << "size.int " << (*m)["line"][42]["size"].int_() << "\n";
    std::cout << "missing " << ((*m)["nope"].truthy() ? "True" : "False") << "\n";
    std::cout << "tree " << canon((*m)["line"][999].tree()) << "\n";

    auto one = g.parse("7.7.7.7 - - [x] \"GET / HTTP/1.1\" 200 5\n", "line");
    std::cout << "rule-parse " << (*one)["status"].str() << "\n";

    auto bad = g.parse("this is not a log line");
    std::cout << "failed-parse " << (bad ? "Match" : "None") << "\n";

    try {
        g.parse_or_throw("this is not a log line");
        std::cout << "diag none\n";
    } catch (const rakupp::ParseError& e) {
        std::cout << "diag line " << e.line << " col " << e.col << " rule " << e.rule << "\n";
    }
    return 0;
}

File: bindings/rust/examples/driver.rs
// The Rust half of the byte-identical gate: the same walk as driver.py,
// through the rakulang crate. Run by grammar-smoke.raku; outputs are
// compared byte for byte.
//
//   RAKUPP_LIB_DIR=$BUILD cargo run --example driver <grammar> <input>

use rakulang::{Error, Grammar, Tree};

fn canon(t: &Tree) -> String {
    match t {
        Tree::List(v) => format!("[{}]", v.iter().map(canon).collect::<Vec<_>>().join(",")),
        Tree::Map(m) => format!( // BTreeMap iterates key-sorted, matching the twins
            "{{{}}}",
            m.iter().map(|(k, v)| format!("{}:{}", k, canon(v))).collect::<Vec<_>>().join(",")),
        Tree::Null => String::new(),
        Tree::Bool(b) => if *b { "True".into() } else { "False".into() },
        Tree::Int(i) => i.to_string(),
        Tree::Num(n) => n.to_string(),
        Tree::Str(s) => s.clone(),
    }
}

fn raku_bool(b: bool) -> &'static str {
    if b { "True" } else { "False" }
}

fn main() -> Result<(), Error> {
    let args: Vec<String> = std::env::args().collect();
    let (grammar_file, input_file) = (&args[1], &args[2]);

    let g = Grammar::from_file(grammar_file, "Log", "LogActions")?;
    let input = std::fs::read_to_string(input_file).expect("input");
    let m = g.parse(&input, "")?.expect("gate: the log corpus did not parse");

    let mut out = String::new();
    let lines = m.get("line");
    let n = lines.len()?;
    out += &format!("lines {}\n", n);
    out += &format!("islist {}\n", raku_bool(lines.is_list()?));

    for i in 0..n {
        out += &format!("{} {}\n", lines.at(i).get("ip").str()?, lines.at(i).get("status").str()?);
    }

    out += &format!("made {}\n", canon(&m.get("line").at(0).get("size").made()?));
    out += &format!("req.str {}\n", m.get("line").at(42).get("req").str()?);
    out += &format!("size.int {}\n", m.get("line").at(42).get("size").int()?);
    out += &format!("missing {}\n", raku_bool(m.get("nope").truthy()?));
    out += &format!("tree {}\n", canon(&m.get("line").at(999).tree()?));

    let one = g.parse("7.7.7.7 - - [x] \"GET / HTTP/1.1\" 200 5\n", "line")?.expect("rule parse");
    out += &format!("rule-parse {}\n", one.get("status").str()?);

    out += &format!("failed-parse {}\n",
                    if g.parse("this is not a log line", "")?.is_some() { "Match" } else { "None" });

    match g.parse_strict("this is not a log line", "") {
        Err(Error::Parse { line, col, rule, .. }) =>
            out += &format!("diag line {} col {} rule {}\n", line, col, rule),
        _ => out += "diag none\n",
    }

    print!("{}", out);
    Ok(())
}
